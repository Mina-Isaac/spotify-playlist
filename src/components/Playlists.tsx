import React, { useEffect } from "react";
import useInfiniteScroll from "../utilities/useInfiniteScroll";
import { observer } from "mobx-react";
import { useMst } from "../tree/Root";
import { Link } from "react-router-dom";

const App = () => {
  const root = useMst();
  const list = root.Playlists.items;
  const setIsFetching = useInfiniteScroll(moreData)[1]
  const loading = root.Playlists.loading;

  function moreData() {
    root.Playlists.fetchNext();
    setIsFetching(false);
  }

  useEffect(() => {
    root.Playlists.fetchNext();
  }, []);

  if (list.length === 0) {
    return <h1>Loading...</h1>;
  }

  return (
    <>
      <h1>List of featured playlists</h1>
      <ul className="list">
        {list.map((item, i) => (
          <li className="card" key={i}>
            <Link
              className="link"
              to={{
                pathname: `/spotify-playlist/artists/${item.id}`,
                state: item.name,
              }}
            >
              <img
                width={"120"}
                src={item.images[0].url}
                alt={item.description}
              />
              <div className="txt">
                <span className="primary-txt">{item.name}</span>
                <span className="secondery-txt">{`by ${item.owner.display_name}`}</span>
              </div>
            </Link>
          </li>
        ))}
      </ul>
      {loading && <span className='primary-text'>Loading...</span> }
    </>
  );
};

export default observer(App);
