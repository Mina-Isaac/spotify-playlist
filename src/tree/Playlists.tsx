import { types, flow, Instance } from "mobx-state-tree";
import appService from "./service";

interface IPlaylist extends Instance<typeof playlist> {}
export interface IPlaylists extends Instance<typeof Playlists> {}

const image = types.model({
  height: types.union(types.number, types.null),
  width: types.union(types.number, types.null),
  url: types.string,
});

const owner = types.model({
  display_name: types.string,
});

const playlist = types.model({
  id: types.identifier,
  name: types.string,
  description: types.string,
  images: types.array(image),
  owner,
});

const Playlists = types
  .model({
    items: types.array(playlist),
    next: types.union(types.string, types.null),
    total: types.number,
    loading: types.boolean,
  })
  .actions((self) => {
    const fetchNext = flow(function* () {
      if (self.next === null) return;
      self.loading = true;
      try {
        const newlyFetched = yield appService.fetchPlaylists(self.next);
        self.next = newlyFetched.next;
        self.loading = false;
        (newlyFetched.items as Array<IPlaylist>).forEach((item) =>
          self.items.push(item)
        );
      } catch (error) {
        console.error("Failed to fetch projects", error);
        self.loading = false;
      }
      return;
    });
    return { fetchNext };
  });

export default Playlists;
