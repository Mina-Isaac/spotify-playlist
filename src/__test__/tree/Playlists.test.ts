import Playlists from "../../tree/Playlists";
import { getSnapshot, onSnapshot, onPatch } from "mobx-state-tree";
import { reaction } from "mobx";

it("should create a model instance", () => {
  const item = Playlists.create({
    items: [
      {
        description: "some text",
        id: "some id",
        name: "some name",
        owner: {
          display_name: "some name",
        },
        images: [{ url: "some url", height: 44, width: null }],
      },
    ],
    loading: false,
    next: "some url",
    total: 13,
  });

  expect(getSnapshot(item)).toMatchSnapshot();
});
