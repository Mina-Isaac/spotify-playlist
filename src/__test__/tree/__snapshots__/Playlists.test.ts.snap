// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a model instance 1`] = `
Object {
  "items": Array [
    Object {
      "description": "some text",
      "id": "some id",
      "images": Array [
        Object {
          "height": 44,
          "url": "some url",
          "width": null,
        },
      ],
      "name": "some name",
      "owner": Object {
        "display_name": "some name",
      },
    },
  ],
  "loading": false,
  "next": "some url",
  "total": 13,
}
`;
